licensing: enforce site admin roles

unknwon opened the pull request:
Non-site admins are not allowed in Free tier.

- All new users are promoted to be side admins in Free tier
- Prevent user creation with a malformed/expired license
- Prevent revoking site admins
- Dotcom: no new users can be promoted to site admin automatically

Unit tests are added and tested end-to-end.

Fixes #14023

ryanslade commented on enterprise/cmd/frontend/internal/licensing/enforcement/users.go:
```diff
@@ -50,3 +58,55 @@ func NewPreCreateUserHook(s licensing.UsersStore) func(context.Context) error {
+	}
+}
+
+// NewPreSetUserIsSiteAdmin returns a PreSetUserIsSiteAdmin closure.
```
It's not clear to me what this function is for

ryanslade replied:
Ah, I see lower down where the interface is defined. Might be useful adding more context here too that mentions when this function will be called

ryanslade commented on internal/db/users.go:
```diff
@@ -29,6 +30,13 @@ type users struct {
+	AfterCreateUser func(ctx context.Context, tx dbutil.DB, user *types.User) error
+	// PreSetUserIsSiteAdmin (if set) is a hook called before promoting/revoking a user to be a
+	// site admin.
+	PreSetUserIsSiteAdmin func(isSiteAdmin bool) error
```
Maybe use `Before` instead of `Pre`?

I think it should be either `Pre` and `Post` or `Before` and `After`

ryanslade reviewed (approved):
LGTM apart from some potential renaming and additional comments

unknwon commented:
@ryanslade Thanks! Addressed feedback in https://github.com/sourcegraph/sourcegraph/pull/15570/commits/49be0722049f746ea726d37bfe59501a3e14777a, merging now.

LMK if you have any post-merge comments.
